terraform/eval_apply.go

```diff
@@ -561,9 +561,22 @@ func (n *EvalApplyProvisioners) apply(ctx EvalContext, provs []*configs.Provisio
 		provisioner := ctx.Provisioner(prov.Type)
 		schema := ctx.ProvisionerSchema(prov.Type)
 
-		forEach, forEachDiags := evaluateResourceForEachExpression(n.ResourceConfig.ForEach, ctx)
-		diags = diags.Append(forEachDiags)
-		keyData := EvalDataForInstanceKey(instanceAddr.Key, forEach)
+		var keyData InstanceKeyEvalData
+		switch {
+		case n.When == configs.ProvisionerWhenDestroy:
+			// if this is a destroy provisioner, we can only know the Key value
+			// itself, which is contained in the address
+			switch key := instanceAddr.Key.(type) {
```

Comment:
To keep this more within the `cty` realm, maybe it would be good to add an `InstanceKey.Value` method that encapsulates the conversion to a suitable `cty.Value`. Then we could use it both here and in `EvalDataForInstanceKey` and thus avoid these two functionally-equivalent-but-different codepaths? :thinking: 

I'd ideally also like to somehow encapsulate the rule that a string key goes in `EachKey` but an int key goes in `CountIndex` in a single place too, but I'm not sure what's a good way to do that. One idea that comes to mind is to just call `EvalDataForInstanceKey` with a nil map, thus letting `EachValue` end up getting populated as `cty.NilVal`, but that feels like relying on implementation details than a properly-designed API. :confounded:

Reply:
Yes, I forgot to actually look at the `EvalDataForInstanceKey` function, as it wasn't doing what I though it was. 

As for the second part, that's what we already do, so I don't see any harm in continuing that at the moment.

Reply:
I liked the InstanceKey.Value idea, and implemented that, along with cleaning up `EvalDataForInstanceKey`. I think it looks a bit cleaner, short of revamping how we handle the InstanceKey data in general.